File: src/test/java/no/nav/syfo/rest/ressurser/MockUtils.java
package no.nav.syfo.rest.ressurser;

import no.nhn.register.fastlegeinformasjon.common.WSArrayOfElectronicAddress;
import no.nhn.register.fastlegeinformasjon.common.WSArrayOfPhysicalAddress;
import no.nhn.schemas.reg.common.en.WSPeriod;
import no.nhn.schemas.reg.flr.*;
import org.mockito.Mockito;
import org.springframework.http.*;
import org.springframework.web.client.RestTemplate;

import javax.xml.datatype.*;
import java.time.LocalDateTime;

import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.when;

public class MockUtils {
    public static final String LEGEKONTOR = "Pontypandy Legekontor";


    public static XMLGregorianCalendar toXMLGregorianCalendar(LocalDateTime localDateTime) {
        XMLGregorianCalendar xmlGregorianCalendar =
                null;
        try {
            xmlGregorianCalendar = DatatypeFactory.newInstance().newXMLGregorianCalendar(localDateTime.toString());
        } catch (DatatypeConfigurationException e) {
            e.printStackTrace();
        }

        return xmlGregorianCalendar;
    }

    public static void mockHarFastlege(IFlrReadOperations fastlegeSoapClient) throws IFlrReadOperationsGetPatientGPDetailsGenericFaultFaultFaultMessage {
        WSArrayOfGPOnContractAssociation fastlege = mockFastlege();
        WSGPOffice legeKontor = mockLegeKontor();
        WSGPContract fastlegeKontrakt = new WSGPContract().withGPOffice(legeKontor);

        WSPatientToGPContractAssociation fastlegeResponse = new WSPatientToGPContractAssociation()
                .withDoctorCycles(fastlege)
                .withGPContract(fastlegeKontrakt)
                .withPeriod(new WSPeriod()
                        .withFrom(toXMLGregorianCalendar(LocalDateTime.now().minusYears(4)))
                        .withTo(toXMLGregorianCalendar(LocalDateTime.now().plusYears(4))))
                .withGPHerId(404);

        Mockito.when(fastlegeSoapClient.getPatientGPDetails(anyString())).thenReturn(fastlegeResponse);
    }

    static void mockIngenFastleger(IFlrReadOperations fastlegeSoapClient) throws IFlrReadOperationsGetPatientGPDetailsGenericFaultFaultFaultMessage {
        Mockito.when(fastlegeSoapClient.getPatientGPDetails(anyString())).thenThrow(new IFlrReadOperationsGetPatientGPDetailsGenericFaultFaultFaultMessage());
    }

    private static WSArrayOfGPOnContractAssociation mockFastlege() {
        return new WSArrayOfGPOnContractAssociation()
                .withGPOnContractAssociations(new WSGPOnContractAssociation()
                        .withHprNumber(123)
                        .withGP(new WSPerson()
                                .withFirstName("Michaela")
                                .withMiddleName("Mike")
                                .withLastName("Quinn")
                                .withNIN("Kake"))
                );
    }

    private static WSGPOffice mockLegeKontor() {
        return new WSGPOffice()
                .withName(LEGEKONTOR)
                .withOrganizationNumber(88888888)
                .withPhysicalAddresses(new WSArrayOfPhysicalAddress())
                .withElectronicAddresses(new WSArrayOfElectronicAddress());
    }

    static void mockResponseFraTilgangskontroll(RestTemplate restTemplate, HttpStatus httpStatus) {
        String okJson = "{\"harTilgang\":true, \"begrunnelse\":\"\"}";
        String forbiddenJson = "{\"harTilgang\":false, \"begrunnelse\":\"GEOGRAFISK\"}";

        ResponseEntity<Object> responseEntity = httpStatus == HttpStatus.FORBIDDEN
                ? new ResponseEntity<>(forbiddenJson, httpStatus)
                : new ResponseEntity<>(okJson, httpStatus);

        when(restTemplate.exchange(
                Mockito.anyString(),
                Mockito.<HttpMethod>eq(HttpMethod.GET),
                Mockito.<HttpEntity<?>>any(),
                Mockito.<Class<Object>>any()
        )).thenReturn(responseEntity);
    }
}

File: src/test/kotlin/testhelper/MockUtils.kt
package testhelper

import no.nhn.register.fastlegeinformasjon.common.WSArrayOfElectronicAddress
import no.nhn.register.fastlegeinformasjon.common.WSArrayOfPhysicalAddress
import no.nhn.schemas.reg.common.en.WSPeriod
import no.nhn.schemas.reg.flr.*
import org.mockito.ArgumentMatchers
import org.mockito.Mockito
import org.springframework.http.*
import org.springframework.web.client.RestTemplate
import java.time.LocalDateTime
import javax.xml.datatype.*

object MockUtils {
    const val LEGEKONTOR = "Pontypandy Legekontor"
    fun toXMLGregorianCalendar(localDateTime: LocalDateTime): XMLGregorianCalendar? {
        var xmlGregorianCalendar: XMLGregorianCalendar? = null
        try {
            xmlGregorianCalendar = DatatypeFactory.newInstance().newXMLGregorianCalendar(localDateTime.toString())
        } catch (e: DatatypeConfigurationException) {
            e.printStackTrace()
        }
        return xmlGregorianCalendar
    }

    @JvmStatic
    @Throws(IFlrReadOperationsGetPatientGPDetailsGenericFaultFaultFaultMessage::class)
    fun mockHarFastlege(fastlegeSoapClient: IFlrReadOperations) {
        val fastlege = mockFastlege()
        val legeKontor = mockLegeKontor()
        val fastlegeKontrakt = WSGPContract().withGPOffice(legeKontor)
        val fastlegeResponse = WSPatientToGPContractAssociation()
            .withDoctorCycles(fastlege)
            .withGPContract(fastlegeKontrakt)
            .withPeriod(WSPeriod()
                .withFrom(LocalDateTime.now().minusYears(4))
                .withTo(LocalDateTime.now().plusYears(4)))
            .withGPHerId(404)
        Mockito.`when`(fastlegeSoapClient.getPatientGPDetails(ArgumentMatchers.anyString())).thenReturn(fastlegeResponse)
    }


    private fun mockFastlege(): WSArrayOfGPOnContractAssociation {
        return WSArrayOfGPOnContractAssociation()
            .withGPOnContractAssociations(WSGPOnContractAssociation()
                .withHprNumber(123)
                .withGP(WSPerson()
                    .withFirstName("Michaela")
                    .withMiddleName("Mike")
                    .withLastName("Quinn")
                    .withNIN("Kake"))
            )
    }

    private fun mockLegeKontor(): WSGPOffice {
        return WSGPOffice()
            .withName(LEGEKONTOR)
            .withOrganizationNumber(88888888)
            .withPhysicalAddresses(WSArrayOfPhysicalAddress())
            .withElectronicAddresses(WSArrayOfElectronicAddress())
    }
}

@Throws(IFlrReadOperationsGetPatientGPDetailsGenericFaultFaultFaultMessage::class)
fun mockIngenFastleger(fastlegeSoapClient: IFlrReadOperations) {
    Mockito.`when`(fastlegeSoapClient.getPatientGPDetails(ArgumentMatchers.anyString())).thenThrow(IFlrReadOperationsGetPatientGPDetailsGenericFaultFaultFaultMessage())
}

fun mockResponseFraTilgangskontroll(restTemplate: RestTemplate, httpStatus: HttpStatus) {
    val okJson = "{\"harTilgang\":true, \"begrunnelse\":\"\"}"
    val forbiddenJson = "{\"harTilgang\":false, \"begrunnelse\":\"GEOGRAFISK\"}"
    val responseEntity = if (httpStatus == HttpStatus.FORBIDDEN) ResponseEntity<Any>(forbiddenJson, httpStatus) else ResponseEntity<Any>(okJson, httpStatus)
    Mockito.`when`(restTemplate.exchange(
        Mockito.anyString(),
        Mockito.eq(HttpMethod.GET),
        Mockito.any(),
        Mockito.any<Class<Any>>()
    )).thenReturn(responseEntity)
}
